kmapper/kmapper.py

```diff
@@ -522,6 +530,36 @@ def map(
             self._summary(graph, str(datetime.now() - start))
 
         return graph
+    
+    def _remove_duplicate_nodes(self, nodes):
+        nodes_to_remove = []
+            
+        node_items = list(nodes.items())
+        num_nodes = len(node_items)
+
+        # nested for-loops over ranges for upper-left triangle of pairwise matrix of node_ids       
```

Comment:
This double for loop approach removed duplicates in O(n^2) time. You should be able to get O(n) by using Python's builtin `Set` object.

Reply:
I think the only way I could make if O(n) is if we dropped the keys of the nodes dict. Arguments for:
* the node key, e.g., "cube1_node1" loses meaning when deduplication happens
* unless I'm mistaken, the node keys don't have much useful information anyway, since it is not clear what dimension they refer to, etc
* dropping the keys would let me make a set of frozensets(node_items), producing unique nodes as output. Could reassign keys something like "node-deduped-1"

Reply:
The only reason I see to keep the node names is so we can remember which hypercube each cluster came from. I have an application now where that is important.

One method could be to create a dictionary with frozensets as the keys and node names as the values. Once everything is added, you can invert the dictionary to the original structure again. 

Another option would be to use a namedTuple (or maybe the new `dataclasses`?) and create a custom `__hash__` function.

Reply:
Given a set of three identical nodes, what key would you want to set for the single retained node after deduplication? My current naive strategy was to retain the first one found -- to always pop the one on the "right" of the nested loop.

Reply:
Hmm I don’t think it matters much. If there are 3, then each hypercube has a common intersection. Maybe it would be better to concatenate each node key so we don’t discard any information. This should be easy with the dictionary approach.